### PR TITLE
Windows Unicode support for ocamlyacc

### PR DESCRIPTION
**Pull request by dra27**
This is part of #1200 which didn't make the merge window for 4.06.0.

The motivation behind this is the elimination of `MKEXE_ANSI`, the presence of which is hurting some work I'm trying to do both extending the FlexDLL bootstrap to the Cygwin ports and extending `configure`'s detection support for FlexDLL/flexlink.

This PR does not alter the way .mly files are interpreted, rather it permits `ocamlyacc` to receive names on the command line which include extended characters. For the most part this is irrelevant, since most resulting names won't be valid module names.

### CODE REVIEW

**Comment by dra27** [comment]
It was apparently a mistake to try to combine this PR with also eliminating `MKEXE_ANSI` from the build system. Assuming this finally builds, I ask for this to be merged first, and then a PR will follow with the additional work for removing `MKEXE_ANSI` completely.

**Review by nojb — COMMENTED** [review]
I did a first pass. Two remarks:

1. It doesn't seem right to use the runtime headers when compiling `ocamlyacc`. It makes the code less self-contained and I wonder how much work are we really saving. As far as I can tell, all we need is a bunch of `_os` macros which could easily be added to `defs.h`.

2. You support compiling both in Unicode and ANSI mode by switching `windows_unicode_enabled`. Is there any reason for this? If not, I would just compile for Unicode unconditionally.

**Comment by dra27** [comment]
@nojb - ocamlyacc was already using the runtime headers, except that before there was duplication between what was in `ocamlyacc` and what was in the runtime (with a non-fatal error where mingw was concerned). I don't think there's any particular reason for ocamlyacc to be kept self-contained, it's just that it happens not to be "original"?

**Comment by dra27** [comment]
I don't have a strong opinion either way on `windows_unicode_enabled` - it was just that in having to lift that function out of the runtime (since it uses OCaml's memory management functions), I either had to hard-code the value or just duplicate the variable. The only reason for doing it this way is that you then don't end up with an ocamlyacc which does recognise UTF8 alongside a compiler which does not... but it's a largely irrelevant distinction. Does it bother you enough to require changing?

**Review by damiendoligez — APPROVED** [review]
Looks OK to me, you may want to get a nod from @nojb before merging.

**Comment by nojb** [comment]
Oups! Very sorry I completely forgot about this one. Yes, I had a few slight questions/remarks about the PR but was convinced by @dra27 answers, so I'm approving as well :)

**Review by nojb — APPROVED** [review]

**Comment by dra27** [comment]
I'll merge once CI gives the nod.

Contrary to @XVilka's excitement in #8797, this PR adds very little of interest for the Windows platform but, as noted above, it does allow me to dust off the revisions to `autoconf` for FlexDLL bootstrapping, so thanks for the approval, @nojb!

**Pull request merged**
